Watchdog API update

fkjagodzinski opened the pull request:
### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->

Bring changes from #10066 to `feature-watchdog` branch. Original description below:

> Watchdog refactoring to multithreaded thread.
-Added API to register muliple threads to watchdog drivers
-Watchdog timeout reconfigures everytime whenever new register thread with longer timeout period
-New APIs for watchdog
wd_register(const osThreadId_t tid, const uint32_t timeout) to register to watchdog
wd_unregister(const osThreadId_t tid) to unregister to watchdog
kick(const osThreadId_t tid) to refresh the watchdog

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->
@rajkan01 @0xc0170 @donatieng @jeromecoutant @jamesbeyond 

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

rajkan01 committed:
Watchdog refactoring to multithreaded thread.

-Added API to register muliple threads to watchdog drivers
-Watchdog timeout reconfigures everytime whenever new register thread with longer timeout period
-New APIs for watchdog
 wd_register(const osThreadId_t tid, const uint32_t timeout) to register to watchdog
 wd_unregister(const osThreadId_t tid) to unregister to watchdog
 kick(const osThreadId_t tid) to refresh the watchdog

rajkan01 committed:
Add HwWatchdog

    - mbed_watchdog_mgr has interface name mbed_wdog_manager_start(),mbed_wdog_manager_stop(),mbed_wdog_manager_kick()
    - HwWatchdog is going to attach with LowPowerTIcker for periodic callback functionality
    - mbed_wdog_manager_start() will either get start either by BL/RTOS Aps,it reads the timeout value specified via macro and macro gets defined in target.json file.
    - mbed_wdog_manager_start() internally configure below HAL hw watchdog with timeout specified via target.json
    - mbed_wdog_manager_start() internally divide the timeout(specified in target.json) by the 2 and attach LowPowerTicker with periodic callback of hw_kick()
    - mbed_wdog_manager_start() internally create one instance of sw watchdog class,to access the static list data structure of sw watchdog class
    - mbed_wdog_manager_kick() function periodically get called and refresh the hw watchdog to avoid watchdog reset
    - converted C++ code into C based APIs
    - added boolean to control watchdog start and stop
    - Added detach from ticker on stop API

rajkan01 committed:
Add SW Watchdog

        -SW watchdog has interface name start(),stop(),kick()	Sw watchdog internally has static list and shared across multiple instance of SW watchdog
	- Sw watchdog initialize timeout value,unique string via constructor whenever threads created sw watchdog object
	-Threads make sure pass proper timeout value,Unique string while creating the instance.
	-start() called by components(BLE,WIFI etc.,),it adds the entry into static list with few details current count ,etc.,
	-kick() called by registered components(BLE,WIFI etc.) to reset current count to zero.
        -is_alive - interface API to mbed_watchdog_manager
        -implementation optimization

rajkan01 committed:
HW Watchdog and SW watchdog interface

  - Added Hw watchdog periodic kick calls software watchdog is_alive check

rajkan01 committed:
Hw Watchdog UnitTest cases

    -Added the unit test case for testing Hw watchdog
    -Added the supported stubs files

rajkan01 committed:
SW watchdog unittest cases

  -Added the mock class function to mock mbed_assert_internal
  -Added the unit test case to test start,kick,stop
  -Modified the interface api name from is_alive to process
  -added the unit test cases for process

fkjagodzinski committed:
Tests: Watchdog: Remove outdated driver tests

The Watchdog driver has been updated making these tests incompatible
with the new API. The Watchdog driver functionality is tested with the
unit tests located in UNITTESTS/drivers/Watchdog/ .

fkjagodzinski committed:
Tests: Watchdog: Add Watchdog Manager tests

rajkan01 committed:
fix: Fixing rebased created bugs

rajkan01 committed:
Fix for the Ci build issue

rajkan01 committed:
fix for spell check and header file

rajkan01 committed:
fix for astyle

rajkan01 committed:
Incorporated the Reivew Comments

rajkan01 committed:
Incorporated the Reivew Comments

rajkan01 committed:
Update Watchdog

 - Changed the process into static method
 - used the singletonptr for creating the low power ticker instance
 - Added the mbed stub into cmake build for cellularnonipsocket,loramacrypto

rajkan01 committed:
Added the ticker support
 -Modified the device dont have support of lp ticker will ticker for watchdog callback register

rajkan01 committed:
Fixed the unittest case failure

rajkan01 committed:
Fix for Problem copying the instance from BSS to stack
  -Created global instance of either LowPowerTicker or Ticker

fkjagodzinski committed:
Tests: ResetReason: Use the HAL watchdog API

The Watchdog driver has been updated making this test incompatible
with the new API. Use HAL API instead.

jamesbeyond reviewed (approved):
The changes already got reviewed, in #10066 , should be good.
We open this new PR only because Raj's on long holiday, we want to get this PR going and merged

fkjagodzinski committed:
Tests: Update Watchdog Manager test

Update the watchdog timing requirements:
1. The watchdog should trigger at, or after the timeout value.
2. The watchdog should trigger before twice the timeout value.
